src/bokeh/io/notebook.py

```diff
@@ -539,6 +581,9 @@ def show_app(app: Application, state: State, notebook_url: str | Callable[[int |
 
     loop = IOLoop.current()
 
+    if os.environ.get("JUPYTER_BOKEH_EXTERNAL_URL"):
+        notebook_url = _remote_jupyter_proxy_url
```

Comment:
If `JUPYTER_BOKEH_EXTERNAL_URL` is set and `notebook_url` is explicitly passed in at the same time, there should probably be a warning issued. I still think a pattern like this might be preferable:
```python
notebook_url = _update_notebook_url_from_env(notebook_url)
```
so that env var handling is encapsulated in one place.

Reply:
I'm looking at this and believe I may have found minor flaws in the type signatures of the parent functions;  in particular I was looking for the default value for notebook_url so I could determine when it is specified.  Type annotations are very new to me but it looks to me like showing.show() and notebook.show_app() alternately omit the callable or provide no default at all.  To address that,  I have defined a type alias ProxyUrl and then in the definitions of show() and show_app() default the value to "localhost:8888".   This is well outside my comfort zone so I figured I should check on this specifically.

Reply:
Here's my type alias:

   ProxyUrl= Union[str, Callable[Union[int, None], str]]

which I use as:

   notebook_url: ProxyUrl = "localhost:8888"

so in one instance I am making up the default,  and the other adding callable (which is already documented in the docstring).

Reply:
@jaytmiller I haven't looked at the type question yet but FYI the relevant codepath goes through here:

https://github.com/bokeh/bokeh/blob/a578a98dc1c4b937cfaf89821c7f184c03b38a36/src/bokeh/io/showing.py#L152

So the default value comes from here: 

https://github.com/bokeh/bokeh/blob/a578a98dc1c4b937cfaf89821c7f184c03b38a36/src/bokeh/io/showing.py#L55

Reply:
Thanks very much for looking at this!  This PR is addressing two entry points,  the one above defines a default but does not permit a callable,  the other one:
https://github.com/bokeh/bokeh/blob/a578a98dc1c4b937cfaf89821c7f184c03b38a36/src/bokeh/io/notebook.py#L501
permits a callable but does not define a default.   That spurred this inquiry because I wasn't sure how to test for "defined" other than to look for "not the default."

One  other thing I just noticed and didn't deal with is this:
https://github.com/bokeh/bokeh/blob/a578a98dc1c4b937cfaf89821c7f184c03b38a36/src/bokeh/io/notebook.py#L164
and the following line:
https://github.com/bokeh/bokeh/blob/a578a98dc1c4b937cfaf89821c7f184c03b38a36/src/bokeh/io/notebook.py#L165
In appears to be a related Protocol but neither mentions the callable nor defines a default.   My version of a fix would replace "str" with "ProxyUrl" and if possible add a default.   I'm wondering however if I should even think about touching a public interface.

Reply:
I would recommend this:
```python
ProxyUrlFunc: TypeAlias = Callable[[int | None], str]
```
And then in all the function signatures (I think, at a quick glance) could then be:
```python
notebook_url: str | ProxyUrlFunc
```

Reply:
> I'm wondering however if I should even think about touching a public interface.

Changing type annotation won't change anything at all for runtime usage of the API.

Reply:
Your suggestion here worked out fine,  thanks!

Reply:
Spoke too soon.   It worked fine as I deployed it,  but I wasn't testing on Python-3.9 and it looks like TypeAlias may be more recent.   What should I do here,  revert it to a simple variable assignment,  or leave it as-is?

Reply:
Have to use 
```
from typing_extensions import TypeAlias
```
for now until we are at minpy=3.11

Reply:
Got it, thanks!